setup.py

```diff
@@ -58,11 +58,13 @@ def main():
                 'futures==3.1.1',
                 'functools32>=3.2,<=3.99'
             ],
-            'pubnub': ['pubnub>=4.0.0,<=4.99']
+            'pubnub': ['pubnub>=4.0.0,<=4.99'],
+            'joblib': ['joblib>=0.11.0,<=0.11.99'],
```

Comment:
I agree with @beckermr that we should try to use as few soft dependencies as possible to make installation more straightforward. `pandas` and `sklearn` seem to me to be good candidates for a soft dependency, while lighter packages (as @mheilman noted) like `requests-toolbelt`, `joblib`, and `pubnub` could be required.

My vote would be to leave it a soft dependency in this PR, but refactor it (along with the other listed packages) to make it required later.  We can then discuss the tradeoffs in that PR.